### PR TITLE
New nice notification modal for errors/success

### PR DESCRIPTION
**Pull request by borbesaur**
Note: I was not able to actually test the changes I made to pKlima because testnet not yet set up.  

## Checklist

- [x] Building the site with `npm run build-site` works without errors
- [x] Building the app with `npm run build-app` works without errors
- [x] I formatted JS and TS files with running `npm run format-all`

### CODE REVIEW

File: app/components/NotificationModal/index.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    dispatch(setAppState({ notificationStatus: null }));
+  };
+
+  return (
+    <>
+      <div className={styles.bg}>
+        <div className={styles.card}>
+          <div className={styles.card_header}>
+            <p>{getAsset("header", status)}</p>
+            <button onClick={closeModal} className={styles.closeButton}>
```
Comment (Atmosfearful) [inline app/components/NotificationModal/index.tsx:75]: Hm the experience is a bit funky when the user closes before the txn is done. I guess its not a dealbreaker.

How do you feel about disabling this button until the final state is reached? success/error

File: app/actions/utils.ts [hunk 2]
```diff
@@ -1,8 +1,29 @@
...
+  if (status && statusType === "error" && message) {
+    if (message === "userRejected")
+      return "You chose to reject the transaction";
+    return message;
+  } else if (statusType === "userConfirmation") {
+    return "Please click 'confirm' in your wallet to continue.";
+  } else if (statusType === "networkConfirmation") {
+    return "Transaction initiated. Waiting for network confirmation.";
+  } else if (statusType === "error") {
+    return "❌ Error: something went wrong...";
```
Comment (Atmosfearful) [inline app/actions/utils.ts:23]: One of mine failed (never seen this error before, probably an RPC hiccup) and gave me this modal:

image

**Review by Atmosfearful — CHANGES_REQUESTED** [review]

**Comment by Atmosfearful** [comment]
@borbesaur also I just tested pklima and it's not triggering the modal anymore. I think maybe the "claimExceeded" state is not being bubbled up properly? if you can't figure it out feel free to pass to me and I can probably figure it out.

File: app/components/views/PKlima/index.tsx [hunk 3]
```diff
@@ -37,7 +40,18 @@ export const PKlima: FC<Props> = (props) => {
   const { provider, address, isConnected } = props;
   const dispatch = useAppDispatch();
 
-  const [status, setStatus] = useState<TxnStatus | "" | "claimExceeded">("");
+  const fullStatus: AppNotificationStatus | null =
+    useSelector(notificationStatus);
+  const status = fullStatus && fullStatus.statusType;
+
+  const setStatus = (status: string, message: string) => {
```
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/components/views/PKlima/index.tsx:47]: i like making a new action creator in appSlice idea

File: app/components/views/Wrap/index.tsx [hunk 4]
```diff
@@ -33,7 +32,17 @@ interface Props {
 export const Wrap: FC<Props> = (props) => {
   const { provider, address, isConnected } = props;
   const dispatch = useAppDispatch();
-  const [status, setStatus] = useState<TxnStatus | "">("");
+  const fullStatus: AppNotificationStatus | null =
+    useSelector(notificationStatus);
+  const status = fullStatus && fullStatus.statusType;
+  const setStatus = (status: string, message: string) => {
+    if (!status) dispatch(setAppState({ notificationStatus: null }));
```
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/components/views/Wrap/index.tsx:39]: ok good point

File: app/actions/pklima.ts [hunk 5]
```diff
@@ -77,64 +76,62 @@ export const changeApprovalTransaction = async (params: {
...
+    params.onStatus("userConfirmation", "");
     const txn = await contract.approve(
       addresses["mainnet"].pklima_exercise,
       value.toString()
     );
-    params.onStatus("networkConfirmation");
+    params.onStatus("networkConfirmation", "");
     await txn.wait(1);
-    params.onStatus("done");
+    params.onStatus("done", "Approval Successfull");
```
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/actions/pklima.ts:86]: good catch.  done.

File: app/components/NotificationModal/index.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    return status.statusType ? modalAssets[section][status.statusType] : "";
+  };
+
+  const closeModal = function (): void {
+    dispatch(setAppState({ notificationStatus: null }));
+  };
+
+  return (
+    <>
+      <div className={styles.bg}>
```
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/components/NotificationModal/index.tsx:71]: also good catch. done.

File: app/components/views/PKlima/index.tsx [hunk 7]
```diff
@@ -2,6 +2,9 @@ import React, { FC, useEffect, useState } from "react";
 import { useSelector } from "react-redux";
 import { providers } from "ethers";
 
+import { notificationStatus } from "state/selectors";
```
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/components/views/PKlima/index.tsx:5]: no problem.  love a good convention to follow.  done.

File: app/components/views/PKlima/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/components/views/PKlima/index.tsx:47]: I honestly did not understand why it was like that so thanks for the explanation

File: app/actions/wrap.ts [hunk 8]
```diff
@@ -16,26 +16,25 @@ export const changeApprovalTransaction = async (params: {
...
+    params.onStatus("userConfirmation", "");
     const txn = await contract.approve(
       addresses["mainnet"].wsklima,
       value.toString()
     );
-    params.onStatus("networkConfirmation");
+    params.onStatus("networkConfirmation", "");
     await txn.wait(1);
-    params.onStatus("done");
+    params.onStatus("done", "Approval successfull");
```
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/actions/wrap.ts:26]: done.

File: app/actions/redeem.ts [hunk 9]
```diff
@@ -68,24 +67,23 @@ export const redeemTransaction = async (params: {
...
-    params.onStatus("userConfirmation");
+    params.onStatus("userConfirmation", "");
     const txn = await contract.migrate(
       ethers.utils.parseUnits(params.value, "ether")
     );
-    params.onStatus("networkConfirmation");
+    params.onStatus("networkConfirmation", "");
     await txn.wait(1);
-    params.onStatus("done");
+    params.onStatus("done", "Bond Redeemed Successfully");
```
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/actions/redeem.ts:76]: good point.  I made them all first word capitalized only now.

File: app/components/views/Wrap/index.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/components/views/Wrap/index.tsx:39]: changed the argument to statusType, not sure why I didn't do that to begin with.

File: app/components/NotificationModal/index.module.css [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+  border: none;
+}
+
+.icon_container {
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  margin: auto;
+  padding: 30px;
+  width: 80px;
```
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/components/NotificationModal/index.module.css:54]: nice ok thanks.  done.

File: app/components/NotificationModal/index.module.css [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+  background: #32972d46;
+  border: 1px solid var(--primary);
+}
+.icon_failure {
+  background: rgba(182, 56, 56, 0.164);
+  border: 1px solid rgba(255, 0, 0, 0.39);
+}
+.icon_confirmation {
+  border: 1px solid var(--primary);
+  border-top-color: var(--surface-01);
```
[the remark this answers is not in the document]
Comment (borbesaur) [reply app/components/NotificationModal/index.module.css:74]: good idea.  nice.  done.

File: app/components/NotificationModal/index.tsx (hunk 1 above)
Comment (borbesaur) [reply app/components/NotificationModal/index.tsx:75]: I'm not a UX designer but I like the idea.  I agree the experience is a little weird if you're like me and always click away every popup as soon as I can.  Plus this way it sort of traps the user so metamask is the only thing they can interact with while the modal is in pending mode and can't accidentally go purchase the same bond 4 times if they don't see the pending status in their metamask.

Comment (borbesaur) [reply app/components/NotificationModal/index.tsx:75]: done.

**Comment by Atmosfearful** [comment]
Looks awesome. Merging this to an experimental branch so I can mess around and try to fix that pklima bug.

**Pull request merged**